Flaky tests log

This log is maintained by LLMs. See our workflow in the repo.

# Flaky Tests Log

Known flaky tests tracked here. Each entry links to the PR where the failure was observed.

## theme.feature

### `theme.feature:34` — Theme preview while navigating palette
### `theme.feature:43` — Theme preview commits on selection

**Failure**: `AssertionError: Expected terminal background #282a36` — `getComputedStyle` returns `rgba(0,0,0,0)` instead of `rgb(40,42,54)` at `theme_steps.ts:47`.

**Platform**: aarch64-darwin only (passes on x86_64-linux).

**Root cause**: Race between theme application and `getComputedStyle` reading the container background. The polling (50 attempts) resolves to the wrong element or reads before the canvas re-renders.

**Observed in**: #321, #322

## sub-terminal.feature

### `sub-terminal.feature:32` — Sub-terminal persists across collapse/expand

**Failure**: `Buffer does not contain "sub-marker" after 50 attempts` — after toggling sub-panel collapse/expand, the sub-terminal buffer is empty (shows fresh prompt instead of previous output).

**Platform**: aarch64-darwin only.

**Root cause**: Buffer polling race after sub-panel expand on slower darwin rendering.

**Observed in**: #321, #322

## sub-terminal.feature

### `sub-terminal.feature:60` — Sub-terminal exit removes tab

**Failure**: `locator.waitFor: Timeout 20000ms exceeded` — waiting for `[data-testid="sub-panel-tab-bar"]` to be hidden after sub-terminal exit.

**Platform**: x86_64-linux (passed on retry).

**Root cause**: Timing race — sub-panel collapse after exit not completing within 20s timeout.

**Observed in**: #322

**Flaky unit test**: `session.test.ts` > `round-trips a saved session`

- **Platform**: x86_64-linux
- **PR**: #554 (commit 2ab15e2)
- **Error**: `AssertionError: expected null not to be null` at `session.test.ts:35`
- **Passed on immediate retry**: yes (71/71)

**Root cause**: Vitest runs 5 server test files in parallel (default `--fileParallelism`), all sharing the same `KOLU_STATE_DIR` (`$$/state` uses the parent shell PID). The `Conf` library's `set()` has a read→update→write gap: another worker reading between T1 (read from disk) and T3 (write back) gets stale data (`null`).

**Fix options**: (1) per-worker state dir via `VITEST_WORKER_ID`, (2) disable file parallelism for the server package.

Two flakes observed on aarch64-darwin during CI for PR #578 (commit fe4edb0):

**Flake 1** — `Canvas-owned scroll does not leak into a terminal` (`features/canvas-mode.feature:95`)
```
Error: xterm received a wheel event — canvas ownership failed to suppress it
  at step_definitions/canvas_mode_steps.ts:367
```
Passed on retry.

**Flake 2** — `Button shows activity when new output arrives while locked` (`features/scroll_lock.feature:32`)
```
AssertionError: Expected button to not be active
'' !== null
  at step_definitions/scroll_lock_steps.ts:146
```
Passed on the next retry.

Both failures are timing-sensitive on the slower aarch64-darwin runner (first one relies on a post-scroll "idle window" expiry, second asserts on a reactive state boundary).

PR #941 (`/do`) — two flaky lanes on HEAD `8a3643b9`:

**`ci::nix@x86_64-linux`** — environmental, GitHub API rate-limit on the ephemeral Incus host:
\`\`\`
error: unable to download 'https://api.github.com/repos/srid/devour-flake/commits/HEAD': HTTP error 403
"API rate limit exceeded for 219.65.110.2"
\`\`\`
Same root cause as prior 403s — unauthenticated fetch on a per-run host. Cascaded all 11 dependent linux recipes to fail.

**`ci::e2e@aarch64-darwin`** — one scenario:
\`\`\`
Scenario: Touch-swiping down inside the terminal scrolls the scrollback up
  features/mobile-terminal-scroll.feature:14
  AssertionError: Expected viewportY to decrease from 0, but it stayed at 118 after 20000ms
    at mobile_terminal_scroll_steps.ts:137
\`\`\`
342 scenarios → 341 passed, 1 failed; failure on retry only. Touch-emulated swipe timing — same shape as prior intermittent mobile-scroll flakes. Same scenario passed on the previous PR-commit run (HEAD `b0838394`).

Retrying both lanes positionally.

## Flake: `ci::nix@x86_64-linux` — pnpm segfault during `kolu-pnpm-deps`

**PR**: #957 (prohibited-keybinds)
**Run**: 2026-05-22, ephemeral container `kolu-pr-957`
**Platform**: `x86_64-linux`

Cascade — every downstream recipe (`ci::biome`, `ci::e2e`, `ci::fmt`, `ci::home-manager`, `ci::install`, `ci::pnpm-hash-fresh`, `ci::smoke`, `ci::surface-example-build`, `ci::typecheck`, `ci::unit`) failed because `ci::nix` is their common upstream.

### Excerpt

```
kolu-pnpm-deps> /nix/store/q2014afk6bp02x3hfz2c94ai251mwzjj-stdenv-linux/setup: line 1806:
  4226 Segmentation fault (core dumped) pnpm install --force --ignore-scripts
       --registry="$NIX_NPM_REGISTRY" --frozen-lockfile
error: Cannot build '/nix/store/xy4gz7f60wl9qj2pzydzscdnx7c8qw4d-kolu-pnpm-deps.drv'.
       Reason: builder failed with exit code 139.
```

Touches no PR code — pure environmental segfault during dependency resolution inside the ephemeral container.

---

## Flake: `ci::unit@aarch64-darwin` — `subscribeGitInfo watcher churn` watcher-count leak

**PR**: #957 (prohibited-keybinds)
**Run**: 2026-05-22, host `sincereintent`
**Platform**: `aarch64-darwin`
**Scenario**: `packages/integrations/git/src/index.test.ts > subscribeGitInfo watcher churn > detects \`git init\` in the current cwd without an OSC 7 setCwd`

### Excerpt

```
AssertionError: expected 1 to be +0 // Object.is equality
- 0
+ 1
 ❯ src/index.test.ts:901:39

  900|   afterEach(() => {
  901|     expect(_sharedHeadWatcherCount()).toBe(0);
                                         ^
  902|     expect(_sharedCwdGitWatcherCount()).toBe(0);
```

Outside the PR scope (file unchanged on `prohibited-keybinds`). The `afterEach` watcher-count assertion suggests a teardown race in the shared HEAD watcher; the `waitFor` timeout above also points at non-deterministic timing in the watcher churn test.

## Flake on PR #960 (`feat/dock-activity-window`, commit `cf08666b`)

Two darwin recipes failed on the initial CI run, passed cleanly on positional retry — classic flake signature.

### `ci::unit@aarch64-darwin` — `packages/integrations/git`

```
× watchGitHead > a HEAD change fans out to every subscriber on the shared watcher
  Error: waitFor: predicate did not become true within 2000ms
  at src/index.test.ts:698:11

× watchGitHead > a listener that throws does not block its peers
  AssertionError: expected 1 to be +0 // Object.is equality
  at src/index.test.ts:724:39
  (`_sharedHeadWatcherCount` non-zero — leaked subscriber after timeout)
```

FSEvents propagation timing — the 2000ms `waitFor` budget on darwin doesn't survive contention. Both passed on retry.

### `ci::typecheck@aarch64-darwin` — `kolu-io@0.1.0`

```
/bin/sh: /Users/srid/.../packages/integrations/io/node_modules/.bin/tsc: No such file or directory
ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  kolu-io@0.1.0 typecheck: `tsc --noEmit`
```

Missing `tsc` binary symlink in `node_modules/.bin` — `pnpm install` interleaving race with another recipe. Passed on retry.

PR: https://github.com/juspay/kolu/pull/960

## Flakes on PR #960 (commit `06d0ec7a`, `feat/dock-activity-window`)

Three flake patterns observed this session, all unrelated to the diff:

### `ci::unit@aarch64-darwin` — git-watcher FSEvents timing

```
× watchGitHead > a HEAD change fans out to every subscriber on the shared watcher
  Error: waitFor: predicate did not become true within 2000ms
```

FSEvents propagation latency on darwin under contention. Reliably reproduced; tight 2000ms `waitFor` budget doesn't survive a busy host.

### `ci::unit@aarch64-darwin` — `terminated by signal 9`

```
error: Recipe `unit` was terminated on line 88 by signal 9
```

Remote `pnpm test:unit` on `sincereintent` killed by SIGKILL mid-run. OOM or external kill; not a test failure. Happens intermittently when multiple recipes run concurrently.

### `ci::e2e@x86_64-linux` (when run locally via `just ci::default`) — mobile-terminal-scroll touch-swipe

```
× Touch-swiping down inside the terminal scrolls the scrollback up (attempt 1, retried)
  AssertionError: Expected viewportY to decrease from 0, but it stayed at 163 after 20000ms
```

Cucumber's built-in retry already kicked in once; both attempts saw viewportY frozen. Touch-gesture dispatch race against xterm.js's pointer-up handler — the swipe `pointerup` may land before xterm sees `pointerdown` propagate. Diff doesn't touch `mobile-terminal-scroll.feature` or its step definitions.

PR: https://github.com/juspay/kolu/pull/960

**Attempt 4** on PR #965 (`b169b013`) — linux ephemeral host (`kolu-pr-965`) provisioned successfully this time. Linux fully green across the board (`ci::e2e@x86_64-linux` + all 10 supporting recipes). Darwin produced three matched-pattern failures:

- `ci::typecheck@aarch64-darwin` — `@kolu/solid-pierre@0.1.0 typecheck: tsc --noEmit` failed via `ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL`. _Same `pnpm install` interleaving race as PR #960's `kolu-io@0.1.0` failure — missing `tsc` binary in `node_modules/.bin` mid-recipe._
- `ci::unit@aarch64-darwin` — `watchGitHead > different repos get independent shared watchers` + `watchGitHead > a fresh subscribe after teardown installs a new watcher`. _Same FSEvents propagation pattern PR #960 hit — 2000ms `waitFor` budget doesn't survive contention on darwin._
- `ci::e2e@aarch64-darwin` — 4 final failures (357 scenarios, 353 passed). Still in the `code-tab` / `file-ref-link` / `claude-code` flake cluster this issue tracks; none in the diff's code path.

Linux `ci::e2e@x86_64-linux` **fully passed** — satisfies the `.agency/do.md` verification standard for PR #965.

**Scenario:** `ci::nix@aarch64-darwin` (and dependent `ci::pnpm-hash-fresh@aarch64-darwin`)
**Platform:** `aarch64-darwin` / `sincereintent`
**PR:** [#989](https://github.com/juspay/kolu/pull/989) (`maximize-lag`)

Infrastructure-level failure, not a code flake — `sincereintent` is out of disk space. 3 consecutive attempts on the same SHA (`690f24b8`) all failed with the same error during the `kolu-website` derivation's pnpm install:

```
ERR_PNPM_ENOSPC  ENOSPC: no space left on device, copyfile
  '/nix/var/nix/builds/nix-94427-1356503230/tmp.dsyjMwnI6h/v10/files/01/...'
  -> '/nix/var/nix/builds/.../kolu-website-src/node_modules/.pnpm/@volar+typescript@2.4.28/...'

tar: ./v10/index: Cannot mkdir: No space left on device
ERROR: pnpm failed to install dependencies
```

Subsequent retries showed the secondary failure mode (tar mkdir failures) — store completely full, no recovery possible without nix-collect-garbage on the host. `x86_64-linux` lane (ephemeral Incus container per .agency/do.md) all 11 recipes green on the same SHA.

Marking #989 as ship-blocked on darwin host cleanup; re-running `ci::nix@aarch64-darwin` should pass once disk space is reclaimed.

**Flake: `ci::unit@x86_64-linux`, `packages/memorable-names` missing vitest binary**

PR: #1007 (commit `073e821a`).
Platform: `x86_64-linux` (ephemeral `kolu-pr-1007` container — fresh `pu create`, never used before).
Retry: passed on first re-run.

`ci::install` succeeded; `pnpm install` printed `Already up to date`. Every package's `vitest run` step worked except `memorable-names`, whose `node_modules/.bin/vitest` symlink was missing at exec time:

```
packages/memorable-names test:unit: /bin/sh: /home/toor/.local/state/justci/073e821/x86_64-linux/src/packages/memorable-names/node_modules/.bin/vitest: No such file or directory
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  memorable-names@0.1.0 test:unit: `vitest run`
spawn ENOENT
```

Smells like a pnpm-store → workspace-package symlink race — pnpm reports "Already up to date" (store satisfies the lockfile) but the per-workspace `.bin/` shim wasn't materialized for `memorable-names` specifically. Other workspaces with `vitest` in `devDependencies` resolved fine in the same install. Re-running `ci::unit@x86_64-linux` against the same container (no `pnpm install` rerun in the recipe DAG) passed; the `.bin/` directory presumably reconciled by the time the second invocation hit it.

**Flake** — `ci::e2e@aarch64-darwin`, PR #1060 (`fix-reconnect-handoff-wedge`).

7 scenarios failed on the darwin lane; the **identical e2e suite passed green on `x86_64-linux`**, and the PR only touches `@kolu/surface` stdio-link internals (nothing to do with the Code tab, file-ref links, git-context, or terminal scrolling). All failures are 20s UI-timeout waits — the darwin-runner-load signature the `e2e` recipe's `CUCUMBER_RETRY=1` safety net is meant to absorb.

Failed scenarios (platform `aarch64-darwin`):
- `features/file-ref-link.feature:28` — Bare filename resolves when its basename is unique in the repo
- `features/file-ref-link.feature:39` — Clicking a slash-containing path opens the file at the line
- `features/file-ref-link.feature:51` — Clicking a bare path (no line number) opens the file with no selection
- `features/file-ref-link.feature:75` — Bare basename without a line number resolves via unique-basename fallback
- `features/file-ref-link.feature:86` — Clicking a line-range file-ref selects the whole range
- `features/git-context.feature:59` — Git context updates when .git appears in cwd without an OSC 7 re-emit
- `features/mobile-terminal-scroll.feature:14` — Touch-swiping down inside the terminal scrolls the scrollback up

Representative error excerpt:
```
✖ And the Code tab should be active # step_definitions/code_tab_steps.ts:257
    locator.waitFor: Timeout 20000ms exceeded.
```
```
AssertionError: Expected viewportY to decrease from 0, but it stayed at 163 after 20000ms
```

Cleared via positional retry: `CI=true nix run github:juspay/ci -- run e2e@aarch64-darwin`.

Recurred on #1063 (R-4 A2), **aarch64-darwin** (`nix-infra@rasam`), commit `4ba8aa9`. The **linux lane's e2e passed fully** (`ci::e2e@x86_64-linux` ✓, 1m34s) and every other darwin node was green; only `ci::e2e@aarch64-darwin` failed, with the usual rasam load-dependent cluster (step-exec time **19m39s**). 8 of 375 scenarios failed — all already auto-retried by justci (`attempt 1, retried`) — the same code-tab / git-context / mobile-scroll cluster:

- `mobile-terminal-scroll.feature:14` — `AssertionError: Expected viewportY to decrease from 0, but it stayed at 163 after 20000ms`
- `git-context.feature:59` ("the header should show a branch name") — `page.waitForFunction: Timeout 20000ms exceeded`
- `code-tab` file-ref range select — `locator('[data-testid="pierre-diff-view"], [data-testid="pierre-file-view"]')` — `locator.waitFor: Timeout 60000ms exceeded`

Unrelated to the PR's diff (a chrome-bar identity rail + additive `server.info` fields). Retrying the darwin lane.

--- a/.agents/skills/parcel/SKILL.md
+++ b/.agents/skills/parcel/SKILL.md
@@ -202,8 +202,26 @@ debounce path swallows event paths silently.
    Rule: **serialize a directory's parcel calls** (kolu does this in
    `sequenceParcelCall`). Pinned by
    `working-tree-watcher.churn.test.ts`, which loads the threadpool to force
-   the lossy ordering.
-5. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
+   the lossy ordering — on linux/inotify only, for the reason below.
+5. **macOS: delivery can run ~15s behind, and parcel's stream is the reason.**
+   Parcel's `startStream` (`src/macos/FSEventsBackend.cc`) creates the stream
+   *without* `kFSEventStreamCreateFlagNoDefer`, so the first event's delivery
+   is deferred rather than immediate. On a volume under constant churn the
+   defer window collapses to the daemon's maximum coalescing window. Measured
+   on kolu's darwin CI box (`ci@petit`, macOS 26.5.2, juspay/kolu#2175): first
+   event **14.2s** after the write, then batches every **15.0s**, 53/53 events
+   delivered — nothing lost, everything late. `fs.watch` in the same process,
+   on the same directory, answered in **1.3s**; libuv passes `NoDefer`. That
+   box's `fseventsd` was also 21 days up, 6.5 GB resident, ~190% CPU — worth
+   checking before concluding anything about a *healthy* mac.
+
+   Two consequences. A test may not gate on a parcel event arriving on darwin
+   (`fs-watch-delivery.testlib.ts` is the one flag that says so; #2175 also
+   records a rarer case where a post-churn subscription delivered nothing at
+   all for 60s+, reproduced with no kolu code in the loop). And on a macOS host
+   whose filesystem is this busy, kolu's own Code tab is that far behind —
+   worth remembering before chasing a "stale git status" report as a kolu bug.
+6. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
    binary. Today the `@parcel/watcher` binary loads its own libstdc++ via
    fallback paths and works, but if a future parcel-watcher version pulls in
    a harder dynamic-link requirement, expect to revisit this.

--- a/.apm/skills/parcel/SKILL.md
+++ b/.apm/skills/parcel/SKILL.md
@@ -202,8 +202,26 @@ debounce path swallows event paths silently.
    Rule: **serialize a directory's parcel calls** (kolu does this in
    `sequenceParcelCall`). Pinned by
    `working-tree-watcher.churn.test.ts`, which loads the threadpool to force
-   the lossy ordering.
-5. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
+   the lossy ordering — on linux/inotify only, for the reason below.
+5. **macOS: delivery can run ~15s behind, and parcel's stream is the reason.**
+   Parcel's `startStream` (`src/macos/FSEventsBackend.cc`) creates the stream
+   *without* `kFSEventStreamCreateFlagNoDefer`, so the first event's delivery
+   is deferred rather than immediate. On a volume under constant churn the
+   defer window collapses to the daemon's maximum coalescing window. Measured
+   on kolu's darwin CI box (`ci@petit`, macOS 26.5.2, juspay/kolu#2175): first
+   event **14.2s** after the write, then batches every **15.0s**, 53/53 events
+   delivered — nothing lost, everything late. `fs.watch` in the same process,
+   on the same directory, answered in **1.3s**; libuv passes `NoDefer`. That
+   box's `fseventsd` was also 21 days up, 6.5 GB resident, ~190% CPU — worth
+   checking before concluding anything about a *healthy* mac.
+
+   Two consequences. A test may not gate on a parcel event arriving on darwin
+   (`fs-watch-delivery.testlib.ts` is the one flag that says so; #2175 also
+   records a rarer case where a post-churn subscription delivered nothing at
+   all for 60s+, reproduced with no kolu code in the loop). And on a macOS host
+   whose filesystem is this busy, kolu's own Code tab is that far behind —
+   worth remembering before chasing a "stale git status" report as a kolu bug.
+6. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
    binary. Today the `@parcel/watcher` binary loads its own libstdc++ via
    fallback paths and works, but if a future parcel-watcher version pulls in
    a harder dynamic-link requirement, expect to revisit this.

--- a/.claude/skills/parcel/SKILL.md
+++ b/.claude/skills/parcel/SKILL.md
@@ -202,8 +202,24 @@ debounce path swallows event paths silently.
    Rule: **serialize a directory's parcel calls** (kolu does this in
    `sequenceParcelCall`). Pinned by
    `working-tree-watcher.churn.test.ts`, which loads the threadpool to force
-   the lossy ordering.
-5. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
+   the lossy ordering — on linux/inotify only, for the reason below.
+5. **macOS: delivery can run ~15s behind, and parcel's stream is the reason.**
+   Parcel's `startStream` (`src/macos/FSEventsBackend.cc`) creates the stream
+   *without* `kFSEventStreamCreateFlagNoDefer`, so the first event's delivery
+   is deferred rather than immediate. On a volume under constant churn the
+   defer window collapses to the daemon's maximum coalescing window. Measured
+   on kolu's darwin CI box (`ci@petit`, macOS 26.5.2, juspay/kolu#2175): first
+   event **14.2s** after the write, then batches every **15.0s**, 53/53 events
+   delivered — nothing lost, everything late. `fs.watch` in the same process,
+   on the same directory, answered in **1.3s**; libuv passes `NoDefer`.
+
+   Two consequences. A test may not gate on a parcel event arriving on darwin
+   (see `fs-watch-delivery.testlib.ts` — it also records a rarer case where a
+   post-churn subscription delivered nothing at all for 60s+, reproduced with
+   no kolu code in the loop). And on a macOS host whose filesystem is this
+   busy, kolu's own Code tab is that far behind — worth remembering before
+   chasing a "stale git status" report as a kolu bug.
+6. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
    binary. Today the `@parcel/watcher` binary loads its own libstdc++ via
    fallback paths and works, but if a future parcel-watcher version pulls in
    a harder dynamic-link requirement, expect to revisit this.

--- a/.claude/skills/parcel/SKILL.md
+++ b/.claude/skills/parcel/SKILL.md
@@ -211,14 +211,16 @@ debounce path swallows event paths silently.
    on kolu's darwin CI box (`ci@petit`, macOS 26.5.2, juspay/kolu#2175): first
    event **14.2s** after the write, then batches every **15.0s**, 53/53 events
    delivered — nothing lost, everything late. `fs.watch` in the same process,
-   on the same directory, answered in **1.3s**; libuv passes `NoDefer`.
+   on the same directory, answered in **1.3s**; libuv passes `NoDefer`. That
+   box's `fseventsd` was also 21 days up, 6.5 GB resident, ~190% CPU — worth
+   checking before concluding anything about a *healthy* mac.
 
    Two consequences. A test may not gate on a parcel event arriving on darwin
-   (see `fs-watch-delivery.testlib.ts` — it also records a rarer case where a
-   post-churn subscription delivered nothing at all for 60s+, reproduced with
-   no kolu code in the loop). And on a macOS host whose filesystem is this
-   busy, kolu's own Code tab is that far behind — worth remembering before
-   chasing a "stale git status" report as a kolu bug.
+   (`fs-watch-delivery.testlib.ts` is the one flag that says so; #2175 also
+   records a rarer case where a post-churn subscription delivered nothing at
+   all for 60s+, reproduced with no kolu code in the loop). And on a macOS host
+   whose filesystem is this busy, kolu's own Code tab is that far behind —
+   worth remembering before chasing a "stale git status" report as a kolu bug.
 6. **`dontFixup = true` in `default.nix`** skips patchELF on the native `.node`
    binary. Today the `@parcel/watcher` binary loads its own libstdc++ via
    fallback paths and works, but if a future parcel-watcher version pulls in

--- a/apm.lock.yaml
+++ b/apm.lock.yaml
@@ -541,7 +541,7 @@ deployments:
   owners:
   - .
   active_owner: .
-  content_hash: sha256:bc3da0d9ff0e372b2701adbcc9a262227b8af21dc1a61022a2b86437af3b5d43
+  content_hash: sha256:a080a71004a542b6c1994ac46fa2a1cb68936f5fc8e49ac6cadb2b9b9770199b
 - kind: project-relative
   target: agents
   value: .agents/skills/perf-diagnose
@@ -1613,7 +1613,7 @@ deployments:
   owners:
   - .
   active_owner: .
-  content_hash: sha256:bc3da0d9ff0e372b2701adbcc9a262227b8af21dc1a61022a2b86437af3b5d43
+  content_hash: sha256:a080a71004a542b6c1994ac46fa2a1cb68936f5fc8e49ac6cadb2b9b9770199b
 - kind: project-relative
   target: claude
   value: .claude/skills/perf-diagnose
@@ -2787,7 +2787,7 @@ local_deployed_file_hashes:
   .agents/skills/evidence/CAPTURE.md: sha256:2fbf0b30b264a91ab156807d1e6d32a4782fcf2c6afd32466ec32a3df3be7ef7
   .agents/skills/evidence/SKILL.md: sha256:8534468a198e753fa9ed6d95477444d418083d92fa7853f1cd6d1c3f93dd1228
   .agents/skills/overnight-ci-flakes/SKILL.md: sha256:f06c0d8399431d777647faf63c3cb9b86e7b1112758e9db00fd86f8322db4b9c
-  .agents/skills/parcel/SKILL.md: sha256:bc3da0d9ff0e372b2701adbcc9a262227b8af21dc1a61022a2b86437af3b5d43
+  .agents/skills/parcel/SKILL.md: sha256:a080a71004a542b6c1994ac46fa2a1cb68936f5fc8e49ac6cadb2b9b9770199b
   .agents/skills/perf-diagnose/SKILL.md: sha256:253d7e174344b2481b2f0a8f0f12ffe206c5f84974bff02b48db297e5023aa0c
   .agents/skills/pierre/SKILL.md: sha256:0c9e0f6c6cb93518bc0fbf5edf22dd57ba0038a35bb6dc757200e0e2327070e8
   .agents/skills/release/SKILL.md: sha256:7451962e4ce6479935d1ffb17013cf5f55c2d5cbe30f0b20e0c3ded533473b28
@@ -2825,7 +2825,7 @@ local_deployed_file_hashes:
   .claude/skills/evidence/CAPTURE.md: sha256:2fbf0b30b264a91ab156807d1e6d32a4782fcf2c6afd32466ec32a3df3be7ef7
   .claude/skills/evidence/SKILL.md: sha256:8534468a198e753fa9ed6d95477444d418083d92fa7853f1cd6d1c3f93dd1228
   .claude/skills/overnight-ci-flakes/SKILL.md: sha256:f06c0d8399431d777647faf63c3cb9b86e7b1112758e9db00fd86f8322db4b9c
-  .claude/skills/parcel/SKILL.md: sha256:bc3da0d9ff0e372b2701adbcc9a262227b8af21dc1a61022a2b86437af3b5d43
+  .claude/skills/parcel/SKILL.md: sha256:a080a71004a542b6c1994ac46fa2a1cb68936f5fc8e49ac6cadb2b9b9770199b
   .claude/skills/perf-diagnose/SKILL.md: sha256:253d7e174344b2481b2f0a8f0f12ffe206c5f84974bff02b48db297e5023aa0c
   .claude/skills/pierre/SKILL.md: sha256:0c9e0f6c6cb93518bc0fbf5edf22dd57ba0038a35bb6dc757200e0e2327070e8
   .claude/skills/release/SKILL.md: sha256:7451962e4ce6479935d1ffb17013cf5f55c2d5cbe30f0b20e0c3ded533473b28

--- a/packages/integrations/git/src/fs-watch-delivery.testlib.ts
+++ b/packages/integrations/git/src/fs-watch-delivery.testlib.ts
@@ -3,45 +3,27 @@
  * filesystem-watch EVENT ARRIVING: darwin does not deliver one on a clock a
  * test can wait out.
  *
- * ## What was measured
+ * WHY the platform behaves this way — parcel's stream omitting
+ * `kFSEventStreamCreateFlagNoDefer`, the daemon's coalescing window, the state
+ * of the box — is parcel knowledge, and lives once, in the `parcel` skill's
+ * failure mode 5. juspay/kolu#2175 carries the full probe transcript. What
+ * belongs HERE is only what decides the skip:
  *
- * On kolu's darwin CI box (`ci@petit`, macOS 26.5.2, aarch64, load ~7),
- * against `@parcel/watcher@2.5.6` with the `fs-events` backend this package
- * pins — the numbers behind juspay/kolu#2175:
+ *   * measured on the darwin CI lane, a live subscription's first event
+ *     arrived **14.2s** after the write and then in 15.0s batches, with
+ *     everything eventually delivered. These cases allow 3s, and
+ *     linux/inotify answers the same writes in milliseconds — so a HEALTHY
+ *     watcher reads as dead there, deterministically;
+ *   * a wider budget does not rescue them. Re-run there with one, three of
+ *     four teardown-and-rebuild cycles landed near 14s and the fourth stayed
+ *     silent for the full 180s — and that silence is not this package's: a
+ *     script driving `@parcel/watcher` directly, no kolu code in the loop,
+ *     reproduced it once in 16 cycles. Widening would trade a deterministic
+ *     red for a flaky one on a REQUIRED status, and pay lane time for it.
  *
- *   * a live subscription delivered its first event **14.2s** after the write,
- *     then in batches every **15.0s**. 53 of 53 events arrived: nothing was
- *     lost, everything was late. Linux/inotify answers the same writes in
- *     milliseconds;
- *   * the churn cases below, re-run there with a widened budget, passed with
- *     deliveries at 12.5s / 13.5s / 14.5s — three of four teardown-and-rebuild
- *     cycles — while the fourth stayed silent for the full **180s**;
- *   * that residual silence is NOT this package: a standalone script driving
- *     `@parcel/watcher` directly, with no kolu code in the loop, ran 16 of the
- *     same subscribe→unsubscribe→subscribe→write cycles and had one deliver
- *     nothing within 60s (deliveries otherwise 0.2s–14.7s);
- *   * `fs.watch` — libuv's own FSEvents stream, same process, same directory —
- *     answered in **1.3s** throughout. The daemon was serving that host
- *     promptly, so the lateness belongs to parcel's stream, which omits
- *     `kFSEventStreamCreateFlagNoDefer` (`src/macos/FSEventsBackend.cc`,
- *     `startStream`). Deferred delivery on a volume under constant churn — a
- *     box building Nix and running e2e all day — collapses to the daemon's
- *     maximum coalescing window, and 15s is what that window measured. The
- *     host's `fseventsd` was 21 days up, 6.5 GB resident, burning ~190% CPU.
- *
- * ## Why these tests are skipped rather than given a bigger budget
- *
- * A wider budget fixes the 15s quantization but not the residual
- * non-delivery, so darwin would trade a deterministic red for a flaky one on
- * a REQUIRED status — the worst of both — and pay tens of seconds of lane
- * time for it. Nothing here is skipped for convenience: the invariants are
- * platform-independent and linux/inotify pins them on every commit, in
- * milliseconds, with the same code under test.
- *
- * Local darwin developers skip these too — a busy laptop produces the same
- * false negatives. juspay/kolu#2175 carries the full probe transcript and the
- * three ways out: get `fseventsd` healthy on the box, an upstream parcel that
- * passes `NoDefer`, or a level-triggered poll beneath the watcher's fast path
- * (the shape `kolu-io`'s `refcounted-dir-watcher` already uses).
+ * Nothing here is skipped for convenience: the invariants are
+ * platform-independent, and linux/inotify pins them on every commit, in
+ * milliseconds, against the same code. Local darwin developers skip them too —
+ * a busy laptop produces the same false negatives.
  */
 export const SKIP_DARWIN_FSWATCH = process.platform === "darwin";

--- a/packages/integrations/git/src/fs-watch-delivery.testlib.ts
+++ b/packages/integrations/git/src/fs-watch-delivery.testlib.ts
@@ -3,27 +3,45 @@
  * filesystem-watch EVENT ARRIVING: darwin does not deliver one on a clock a
  * test can wait out.
  *
+ * Two different watch paths land under this flag, and they fail differently —
+ * both measured on the darwin CI box, neither answerable by a bigger budget.
+ *
+ * ## `@parcel/watcher` — late, but consistently so
+ *
+ * The working-tree churn cases. A live subscription's first event arrived
+ * **14.2s** after the write and then in 15.0s batches, everything eventually
+ * delivered; those cases allow 3s, so a HEALTHY watcher reads as dead,
+ * deterministically. Widening does not rescue them: re-run there with a wide
+ * budget, three of four teardown-and-rebuild cycles landed near 14s and the
+ * fourth stayed silent for the full 180s — and that silence is not this
+ * package's, since a script driving `@parcel/watcher` directly, with no kolu
+ * code in the loop, reproduced it once in 16 cycles. So darwin would trade a
+ * deterministic red for a flaky one on a REQUIRED status.
+ *
+ * ## `fs.watch` — not late, unpredictable
+ *
+ * The git-dir watcher cases (HEAD/reflog/index, via `kolu-io`'s
+ * `refcounted-dir-watcher`). Do NOT read the parcel numbers above as this
+ * path's: libuv's own FSEvents stream, probed on the same box in the same
+ * process, answered in **1.3s** — and in a later probe delivered nothing at
+ * all for a 20s window. That spread IS the failure: not a clock to wait out
+ * but non-determinism under contention, which is what these skips were
+ * originally written for (juspay/kolu#320, since closed) and why the file's
+ * `waitForHeadEvent` re-touches HEAD across six attempts rather than trusting
+ * one budget. A test cannot pick a number against that.
+ *
+ * ## Why skipping is honest here
+ *
+ * Nothing is skipped for convenience. Both invariants are
+ * platform-independent — the parcel serialization fix, and the dispatcher's
+ * snapshot + try/catch per listener — and linux/inotify pins both on every
+ * commit, in milliseconds, against the same code. Local darwin developers skip
+ * them too: a busy laptop produces the same false negatives.
+ *
  * WHY the platform behaves this way — parcel's stream omitting
  * `kFSEventStreamCreateFlagNoDefer`, the daemon's coalescing window, the state
- * of the box — is parcel knowledge, and lives once, in the `parcel` skill's
- * failure mode 5. juspay/kolu#2175 carries the full probe transcript. What
- * belongs HERE is only what decides the skip:
- *
- *   * measured on the darwin CI lane, a live subscription's first event
- *     arrived **14.2s** after the write and then in 15.0s batches, with
- *     everything eventually delivered. These cases allow 3s, and
- *     linux/inotify answers the same writes in milliseconds — so a HEALTHY
- *     watcher reads as dead there, deterministically;
- *   * a wider budget does not rescue them. Re-run there with one, three of
- *     four teardown-and-rebuild cycles landed near 14s and the fourth stayed
- *     silent for the full 180s — and that silence is not this package's: a
- *     script driving `@parcel/watcher` directly, no kolu code in the loop,
- *     reproduced it once in 16 cycles. Widening would trade a deterministic
- *     red for a flaky one on a REQUIRED status, and pay lane time for it.
- *
- * Nothing here is skipped for convenience: the invariants are
- * platform-independent, and linux/inotify pins them on every commit, in
- * milliseconds, against the same code. Local darwin developers skip them too —
- * a busy laptop produces the same false negatives.
+ * of that box — is parcel knowledge, and lives once, in the `parcel` skill's
+ * *"macOS: delivery can run ~15s behind"* failure mode. juspay/kolu#2175
+ * carries the full probe transcript.
  */
 export const SKIP_DARWIN_FSWATCH = process.platform === "darwin";

--- a/packages/integrations/git/src/fs-watch-delivery.testlib.ts
+++ b/packages/integrations/git/src/fs-watch-delivery.testlib.ts
@@ -1,0 +1,47 @@
+/**
+ * One platform fact, shared by every test in this package that gates on a
+ * filesystem-watch EVENT ARRIVING: darwin does not deliver one on a clock a
+ * test can wait out.
+ *
+ * ## What was measured
+ *
+ * On kolu's darwin CI box (`ci@petit`, macOS 26.5.2, aarch64, load ~7),
+ * against `@parcel/watcher@2.5.6` with the `fs-events` backend this package
+ * pins — the numbers behind juspay/kolu#2175:
+ *
+ *   * a live subscription delivered its first event **14.2s** after the write,
+ *     then in batches every **15.0s**. 53 of 53 events arrived: nothing was
+ *     lost, everything was late. Linux/inotify answers the same writes in
+ *     milliseconds;
+ *   * the churn cases below, re-run there with a widened budget, passed with
+ *     deliveries at 12.5s / 13.5s / 14.5s — three of four teardown-and-rebuild
+ *     cycles — while the fourth stayed silent for the full **180s**;
+ *   * that residual silence is NOT this package: a standalone script driving
+ *     `@parcel/watcher` directly, with no kolu code in the loop, ran 16 of the
+ *     same subscribe→unsubscribe→subscribe→write cycles and had one deliver
+ *     nothing within 60s (deliveries otherwise 0.2s–14.7s);
+ *   * `fs.watch` — libuv's own FSEvents stream, same process, same directory —
+ *     answered in **1.3s** throughout. The daemon was serving that host
+ *     promptly, so the lateness belongs to parcel's stream, which omits
+ *     `kFSEventStreamCreateFlagNoDefer` (`src/macos/FSEventsBackend.cc`,
+ *     `startStream`). Deferred delivery on a volume under constant churn — a
+ *     box building Nix and running e2e all day — collapses to the daemon's
+ *     maximum coalescing window, and 15s is what that window measured. The
+ *     host's `fseventsd` was 21 days up, 6.5 GB resident, burning ~190% CPU.
+ *
+ * ## Why these tests are skipped rather than given a bigger budget
+ *
+ * A wider budget fixes the 15s quantization but not the residual
+ * non-delivery, so darwin would trade a deterministic red for a flaky one on
+ * a REQUIRED status — the worst of both — and pay tens of seconds of lane
+ * time for it. Nothing here is skipped for convenience: the invariants are
+ * platform-independent and linux/inotify pins them on every commit, in
+ * milliseconds, with the same code under test.
+ *
+ * Local darwin developers skip these too — a busy laptop produces the same
+ * false negatives. juspay/kolu#2175 carries the full probe transcript and the
+ * three ways out: get `fseventsd` healthy on the box, an upstream parcel that
+ * passes `NoDefer`, or a level-triggered poll beneath the watcher's fast path
+ * (the shape `kolu-io`'s `refcounted-dir-watcher` already uses).
+ */
+export const SKIP_DARWIN_FSWATCH = process.platform === "darwin";

--- a/packages/integrations/git/src/index.test.ts
+++ b/packages/integrations/git/src/index.test.ts
@@ -3,18 +3,6 @@ import os from "node:os";
 import path from "node:path";
 import { simpleGit } from "simple-git";
 
-/** Tests in this file that gate on `fs.watch` event delivery are
- *  unreliable on darwin — FSEvents coalesces and can take >12s to
- *  deliver a single change under contention. The dispatcher logic
- *  itself (snapshot + try/catch per listener, in
- *  `kolu-io/refcounted-dir-watcher.ts:96-106`) is verified by
- *  linux+inotify CI on every commit; the darwin skips here only avoid
- *  the platform layer's non-determinism. Local darwin devs also skip
- *  these — running them on a busy laptop produces false negatives.
- *  Tracked: juspay/kolu#320 for a proper fix (test seam or polling
- *  fallback in the production watcher path). */
-const SKIP_DARWIN_FSWATCH = process.platform === "darwin";
-
 import {
   afterAll,
   afterEach,
@@ -34,6 +22,12 @@ import {
   _settledSharedCwdGitWatchers,
   _sharedCwdGitWatcherCount,
 } from "./cwd-git-watcher.ts";
+/** The skips in this file gate on `fs.watch` event delivery, which darwin
+ *  cannot promise inside a test budget — the shared constant carries the
+ *  measurements. The dispatcher logic itself (snapshot + try/catch per
+ *  listener, in `kolu-io/refcounted-dir-watcher.ts`) is verified by
+ *  linux+inotify CI on every commit. */
+import { SKIP_DARWIN_FSWATCH } from "./fs-watch-delivery.testlib.ts";
 import { WATCHER_DEBOUNCE_MS } from "./git-dir.ts";
 import {
   _resetSharedHeadWatchers,

--- a/packages/integrations/git/src/index.test.ts
+++ b/packages/integrations/git/src/index.test.ts
@@ -22,11 +22,6 @@ import {
   _settledSharedCwdGitWatchers,
   _sharedCwdGitWatcherCount,
 } from "./cwd-git-watcher.ts";
-/** The skips in this file gate on `fs.watch` event delivery, which darwin
- *  cannot promise inside a test budget — the shared constant carries the
- *  measurements. The dispatcher logic itself (snapshot + try/catch per
- *  listener, in `kolu-io/refcounted-dir-watcher.ts`) is verified by
- *  linux+inotify CI on every commit. */
 import { SKIP_DARWIN_FSWATCH } from "./fs-watch-delivery.testlib.ts";
 import { WATCHER_DEBOUNCE_MS } from "./git-dir.ts";
 import {

--- a/packages/integrations/git/src/working-tree-watcher.churn.test.ts
+++ b/packages/integrations/git/src/working-tree-watcher.churn.test.ts
@@ -37,6 +37,9 @@
  *
  * The invariant this pins: after a full teardown + rebuild of one repo's shared
  * watcher, a write STILL reaches the new listener.
+ *
+ * Both cases gate on an event ARRIVING, so both run on linux/inotify only —
+ * see `SKIP_DARWIN_FSWATCH` for the measurements that put darwin out of reach.
  */
 
 import crypto from "node:crypto";
@@ -45,6 +48,7 @@ import os from "node:os";
 import path from "node:path";
 import { simpleGit } from "simple-git";
 import { afterEach, beforeEach, describe, expect, it } from "vitest";
+import { SKIP_DARWIN_FSWATCH } from "./fs-watch-delivery.testlib.ts";
 import { watchWorkingTree } from "./working-tree-watcher.ts";
 
 const sleep = (ms: number): Promise<void> =>
@@ -86,7 +90,7 @@ async function subscribeInstalled(
   return off;
 }
 
-describe("watchWorkingTree rebuild", () => {
+describe.skipIf(SKIP_DARWIN_FSWATCH)("watchWorkingTree rebuild", () => {
   let repo: string;
 
   beforeEach(async () => {

--- a/packages/padi/src/memorySampler.test.ts
+++ b/packages/padi/src/memorySampler.test.ts
@@ -1,6 +1,9 @@
 import { Effect } from "effect";
 import { afterEach, describe, expect, it, vi } from "vitest";
-import { withOsfactsMemoryFixture } from "./memorySampler.testlib.ts";
+import {
+  PAUSED_FIXTURE_TIMEOUT_MS,
+  withOsfactsMemoryFixture,
+} from "./memorySampler.testlib.ts";
 
 const endpoint = vi.hoisted(() => ({
   target: undefined as { pid: number; startedAt: number } | undefined,
@@ -129,42 +132,50 @@ describe("samplePadiMemory — osfacts V2 RSS", () => {
     );
   });
 
-  it("does not publish a prior kaval generation after an in-flight recycle", async () => {
-    connectedKaval();
-    await withOsfactsMemoryFixture(
-      {
-        rows: [`M\t${process.pid}\t10485760`, "M\t4242\t20971520"],
-        paused: true,
-      },
-      async (fixture) => {
-        const sample = Effect.runPromise(samplePadiMemory);
-        await fixture.awaitStarted();
-        endpoint.target = { pid: 4242, startedAt: 2_000 };
-        fixture.release();
+  it(
+    "does not publish a prior kaval generation after an in-flight recycle",
+    async () => {
+      connectedKaval();
+      await withOsfactsMemoryFixture(
+        {
+          rows: [`M\t${process.pid}\t10485760`, "M\t4242\t20971520"],
+          paused: true,
+        },
+        async (fixture) => {
+          const sample = Effect.runPromise(samplePadiMemory);
+          await fixture.awaitStarted();
+          endpoint.target = { pid: 4242, startedAt: 2_000 };
+          fixture.release();
 
-        await expect(sample).resolves.toEqual({
-          padi: { status: "ok", rssBytes: 10_485_760 },
-          kaval: { status: "absent" },
-        });
-      },
-    );
-  });
+          await expect(sample).resolves.toEqual({
+            padi: { status: "ok", rssBytes: 10_485_760 },
+            kaval: { status: "absent" },
+          });
+        },
+      );
+    },
+    PAUSED_FIXTURE_TIMEOUT_MS,
+  );
 
-  it("does not misreport a failed prior-generation read as a current kaval error", async () => {
-    connectedKaval();
-    await withOsfactsMemoryFixture(
-      { rows: [], version: 999, paused: true },
-      async (fixture) => {
-        const sample = Effect.runPromise(samplePadiMemory);
-        await fixture.awaitStarted();
-        endpoint.target = { pid: 4242, startedAt: 2_000 };
-        fixture.release();
+  it(
+    "does not misreport a failed prior-generation read as a current kaval error",
+    async () => {
+      connectedKaval();
+      await withOsfactsMemoryFixture(
+        { rows: [], version: 999, paused: true },
+        async (fixture) => {
+          const sample = Effect.runPromise(samplePadiMemory);
+          await fixture.awaitStarted();
+          endpoint.target = { pid: 4242, startedAt: 2_000 };
+          fixture.release();
 
-        await expect(sample).resolves.toEqual({
-          padi: { status: "error" },
-          kaval: { status: "absent" },
-        });
-      },
-    );
-  });
+          await expect(sample).resolves.toEqual({
+            padi: { status: "error" },
+            kaval: { status: "absent" },
+          });
+        },
+      );
+    },
+    PAUSED_FIXTURE_TIMEOUT_MS,
+  );
 });

--- a/packages/padi/src/memorySampler.test.ts
+++ b/packages/padi/src/memorySampler.test.ts
@@ -138,7 +138,7 @@ describe("samplePadiMemory — osfacts V2 RSS", () => {
       },
       async (fixture) => {
         const sample = Effect.runPromise(samplePadiMemory);
-        await vi.waitFor(() => expect(fixture.hasStarted()).toBe(true));
+        await fixture.awaitStarted();
         endpoint.target = { pid: 4242, startedAt: 2_000 };
         fixture.release();
 
@@ -156,7 +156,7 @@ describe("samplePadiMemory — osfacts V2 RSS", () => {
       { rows: [], version: 999, paused: true },
       async (fixture) => {
         const sample = Effect.runPromise(samplePadiMemory);
-        await vi.waitFor(() => expect(fixture.hasStarted()).toBe(true));
+        await fixture.awaitStarted();
         endpoint.target = { pid: 4242, startedAt: 2_000 };
         fixture.release();
 

--- a/packages/padi/src/memorySampler.testlib.ts
+++ b/packages/padi/src/memorySampler.testlib.ts
@@ -19,8 +19,21 @@ import { join } from "node:path";
 import { shellQuoteArg } from "@kolu/shell-quote";
 import { vi } from "vitest";
 
+/** How long the fake osfacts may take to actually be RUNNING.
+ *
+ *  Starting it is a real process launch — `node_modules`-free `/bin/sh`, but a
+ *  launch — and this fixture is used by the daemon-gated suites, which fork
+ *  real kaval/padi daemons and PTYs alongside it. vitest's `waitFor` default is
+ *  1s, which measures the box rather than the sampler: on the darwin CI lane
+ *  the file's ten cases take 1.6–2.6s in total when healthy, and one loaded run
+ *  missed the 1s launch window and reddened `ci::daemon@aarch64-darwin`
+ *  (juspay/kolu#2176). A spawn that has not happened in 30s is a real hang;
+ *  anything short of that is the machine being busy. */
+const SPAWN_BUDGET_MS = 30_000;
+
 export interface OsfactsMemoryFixture {
-  readonly hasStarted: () => boolean;
+  /** Resolve once the fake osfacts has started and is holding at the pause. */
+  readonly awaitStarted: () => Promise<void>;
   readonly readArgs: () => string;
   readonly release: () => void;
 }
@@ -60,17 +73,27 @@ function installOsfactsMemoryFixture(
   process.env.KOLU_OSFACTS_BIN = bin;
   const release = () => writeFileSync(releaseFile, "");
   return {
-    hasStarted: () => existsSync(startedFile),
+    awaitStarted: () =>
+      vi.waitFor(
+        () => {
+          if (!existsSync(startedFile))
+            throw new Error("osfacts fixture has not started");
+        },
+        { timeout: SPAWN_BUDGET_MS, interval: 25 },
+      ),
     readArgs: () => readFileSync(argsFile, "utf8").trim(),
     release,
     restore: async () => {
       try {
         if (existsSync(startedFile) && !existsSync(finishedFile)) {
           release();
-          await vi.waitFor(() => {
-            if (!existsSync(finishedFile))
-              throw new Error("osfacts still running");
-          });
+          await vi.waitFor(
+            () => {
+              if (!existsSync(finishedFile))
+                throw new Error("osfacts still running");
+            },
+            { timeout: SPAWN_BUDGET_MS, interval: 25 },
+          );
         }
       } finally {
         if (previous === undefined) delete process.env.KOLU_OSFACTS_BIN;

--- a/packages/padi/src/memorySampler.testlib.ts
+++ b/packages/padi/src/memorySampler.testlib.ts
@@ -31,6 +31,16 @@ import { vi } from "vitest";
  *  anything short of that is the machine being busy. */
 const SPAWN_BUDGET_MS = 30_000;
 
+/** What a case using a PAUSED fixture must pass as its own timeout.
+ *
+ *  `awaitStarted()`'s budget is only real if the enclosing case outlives it,
+ *  and this package sets no `testTimeout`, so vitest's 5s default would kill
+ *  the case first — capping the wait at 5s and reporting a hang exactly where
+ *  the budget above exists to say "the box is busy". Deriving the case timeout
+ *  from the budget keeps the two from drifting apart again; the headroom is
+ *  because equal-length timers race. */
+export const PAUSED_FIXTURE_TIMEOUT_MS = SPAWN_BUDGET_MS + 5_000;
+
 export interface OsfactsMemoryFixture {
   /** Resolve once the fake osfacts has started and is holding at the pause. */
   readonly awaitStarted: () => Promise<void>;

--- a/packages/padi/src/memorySampler.testlib.ts
+++ b/packages/padi/src/memorySampler.testlib.ts
@@ -22,11 +22,13 @@ import { vi } from "vitest";
 /** How long the fake osfacts may take to actually be RUNNING.
  *
  *  Starting it is a real process launch — `node_modules`-free `/bin/sh`, but a
- *  launch — and this fixture is used by the daemon-gated suites, which fork
- *  real kaval/padi daemons and PTYs alongside it. vitest's `waitFor` default is
- *  1s, which measures the box rather than the sampler: on the darwin CI lane
- *  the file's ten cases take 1.6–2.6s in total when healthy, and one loaded run
- *  missed the 1s launch window and reddened `ci::daemon@aarch64-darwin`
+ *  launch — and vitest's `waitFor` default of 1s measures the box rather than
+ *  the sampler. Nothing in this package runs beside it (`fileParallelism:
+ *  false`, and `test-daemon` passes `--workspace-concurrency=1`), but the CI
+ *  lane runs sibling pipeline recipes on the same host, so a launch competes
+ *  with whatever nix build or e2e run is live there. On the darwin lane the
+ *  file's ten cases take 1.6–2.6s in total when healthy; one loaded run missed
+ *  the 1s launch window and reddened `ci::daemon@aarch64-darwin`
  *  (juspay/kolu#2176). A spawn that has not happened in 30s is a real hang;
  *  anything short of that is the machine being busy. */
 const SPAWN_BUDGET_MS = 30_000;
@@ -51,6 +53,11 @@ export interface OsfactsMemoryFixture {
 export interface OsfactsMemoryFixtureSpec {
   readonly rows: readonly string[];
   readonly version?: number;
+  /** Hold the fake osfacts at a barrier until `release()`, so a case can mutate
+   *  state mid-read. A case that sets this MUST pass
+   *  {@link PAUSED_FIXTURE_TIMEOUT_MS} as its own timeout — it is the only
+   *  shape that waits on a spawn, and vitest's 5s default would otherwise kill
+   *  the case before `awaitStarted()`'s budget is up. */
   readonly paused?: boolean;
 }
 

